add support for tsa signing and verification of images

hectorj2f opened the pull request:
Signed-off-by: Hector Fernandez <hector@chainguard.dev>

<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Pleases use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

In relation to https://github.com/sigstore/cosign/issues/2331, this PR adds support for image signing and verification using a RFC3161 time-stamping server **instead** of the transparency log.

NOTE: Rekor and TSA cannot be used together for the moment.

For testing purposes, we have integrated using https://github.com/sigstore/timestamp-authority. So we expect you to run a TSA server for testing:

```
$ curl http://localhost:3000/api/v1/timestamp/certchain > ts_chain.pem

$ cosign sign <YOUR_IMAGE> --tlog-upload=false --timestamp-server-url=http://localhost:3000

$ cosign verify <YOUR_IMAGE> --timestamp-server-url=http://localhost:3000 --timestamp-cert-chain=ts_chain.pem
```


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->
Add support for signing images using a RFC3161 time-stamping server

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

hectorj2f commented:
@priyawadhwa @haydentherapper I am considering to rename any reference of TSABundle to 3161Timestamp. Any thoughts ?.

hectorj2f commented on pkg/cosign/verify.go:
```diff
@@ -928,6 +946,43 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, rekorClient *client.Re
+		return false, fmt.Errorf("reading base64signature: %w", err)
+	}
+
+	fmt.Println("Verifying TSA Bundle")
```
Perhaps we want to avoid printing this message (it will be printed per validated signature).

znewman01 replied:
If you have it, it should go to STDERR

hectorj2f commented on pkg/cosign/verify.go:
```diff
@@ -928,6 +946,43 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, rekorClient *client.Re
+
+	err = tsaverification.VerifyTimestampResponse(bundle.SignedRFC3161Timestamp, bytes.NewReader(sigBytes), certPool)
+	if err != nil {
+		return false, fmt.Errorf("unable verifyTSRWithPEM: %w", err)
```
I need to fix this error msg.

haydentherapper commented on cmd/cosign/cli/sign/sign.go:
```diff
@@ -231,7 +234,16 @@ func signDigest(ctx context.Context, digest name.Digest, payload []byte, ko opti
-	if ShouldUploadToTlog(ctx, ko, digest, ko.SkipConfirmation, tlogUpload) {
+
+	// TODO: For the moment you can only use the timestamped service OR the transparency log.
+	if ko.TSAServerURL != "" {
```
I would prefer to support both and not have this as a TODO. Timestamps are simply a way to have a third-party provide time, but don't provide transparency. Having this as a TODO risks changing the security model of Sigstore.

Following the existing code, can we have the `tsa.NewSigner` be wrapped by `irekor.NewSigner`?

cc @asraa

hectorj2f replied:
@haydentherapper @asraa Could we add support for both as part of a follow-up PR ? Likewise I will advocate for supporting the TSA as an alternative to the tlog. Some users may not want transparency.

haydentherapper replied:
As long as the follow up is before we cut a new release, sure. Is it a large change to support both? I would have thought we could support both by just wrapping the tsa signer in the rekor signer struct.

If a user doesn't want to use the transparency log, that should be controlled via an explicit flag (`tlog-upload`) rather than the usage of timestamping. I just want to make sure we're clear on the messaging and security model for Sigstore, that timestamping provides an alternative to how to verify expiration of a certificate. Rekor currently plays two roles, transparency and timestamping, so using a TSA only changes the second. 

There should be a limited number of reasons to not use Rekor - private artifacts are the main one. Otherwise, you really should be uploading entries to Rekor.

priyawadhwa replied:
> There should be a limited number of reasons to not use Rekor - private artifacts are the main one. Otherwise, you really should be uploading entries to Rekor.

+1. We should still support this use case though. We can default to using the TSA+Rekor together, and provide an opt-in flag for folks who want to sign private artifacts and verify against the TSA but not Rekor.  I think because of this use case it's more logical to have a separate TSA signer/verifier than to wrap it into Rekor. 

> Could we add support for both as part of a follow-up PR ?

sounds good! as a reviewer I definitely prefer lots of smaller PRs :) there's no plan to cut any releases at the moment so should be fine.

asraa replied:
This all sounds good to me: I think everyone laid it out right. We can support TSA+Rekor, with a control knob on removing hte Rekor upload.

For verification, we can start verifying TSA's on conditional requirements, soon always validating them in the long tail. The verification flow will look the same, we still will use a Rekor promise of inclusion offline, but instead use the verified TSA as the time check.

hectorj2f replied:
I extracted the following list of action items as result of this thread:

* Add support for **TSA and Rekor together**.
* Users could rely on **TSA-only** whenever they set `--tlog-upload=false`. We'll add a warning message and documentation about the limitations of using the timestamp verification **instead** of the tlog verification.
* Add a **`--skip-tlog-verify`** flag to our **`cosign verify*`** commands to skip the tlog verification done by default.

haydentherapper commented on cmd/cosign/cli/options/verify.go:
```diff
@@ -76,6 +79,12 @@ func (o *VerifyOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&o.TSAServerURL, "timestamp-server-url", "",
+		"url to a timestamp RFC3161 server, default none")
+
+	cmd.Flags().StringVar(&o.TSACertChainPath, "timestamp-cert-chain", "",
```
Can you clarify that this should contain the entire chain from the TSA certificate leaf at index 0 to the root at index n

haydentherapper commented on internal/pkg/cosign/tsa/signer.go:
```diff
@@ -0,0 +1,137 @@
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("Calling TSA authority ...")
```
Remove?

haydentherapper commented on internal/pkg/cosign/tsa/signer.go:
```diff
@@ -0,0 +1,137 @@
+	}
+	fmt.Println("Calling TSA authority ...")
+	params := ts.NewGetTimestampResponseParams()
+	params.SetTimeout(time.Second * 30)
```
Can we reduce this, this seems like a long timeout?

hectorj2f replied:
I believe this is the default timeout value used by the timestamp-cli.

haydentherapper replied:
We should update that too. For the sigstore TSA we'll aim for <1s p99s, for third parties, <10 or <5 should be reasonable

haydentherapper commented on internal/pkg/cosign/tsa/signer.go:
```diff
@@ -0,0 +1,137 @@
+		return nil, err
+	}
+
+	fmt.Fprintln(os.Stderr, "Timestamp authority entry created with time:", ts.Time)
```
```suggestion
	fmt.Fprintln(os.Stderr, "Timestamp fetched with time: ", ts.Time)
```

haydentherapper commented on internal/pkg/cosign/tsa/signer.go:
```diff
@@ -0,0 +1,137 @@
+func createTimestampAuthorityRequest(artifactBytes []byte, hash crypto.Hash, policyStr string) ([]byte, error) {
+	reqOpts := &timestamp.RequestOptions{
+		Hash:         hash,
+		Certificates: true, // if the timestamp response should contain a certificate chain
```
```suggestion
		Certificates: true, // if the timestamp response should contain the timestamp certificate
```

haydentherapper replied:
small nit, the chain is not included, only the leaf

haydentherapper commented on internal/pkg/cosign/tsa/signer_test.go:
```diff
@@ -0,0 +1,95 @@
+	return s
+}
+
+func TestSigner(t *testing.T) {
```
Nice tests!

hectorj2f replied:
Thanks 🦾

haydentherapper commented on pkg/cosign/verify.go:
```diff
@@ -928,6 +946,43 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, rekorClient *client.Re
+		return false, fmt.Errorf("unable verifyTSRWithPEM: %w", err)
+	}
+
+	return true, nil
```
We also need to verify the certificate expiration as part of verification, like what we do in the rekor bundle - https://cs.github.com/sigstore/cosign/blob/44a7117e0cee495a163c92e9f347efb2d9b747c7/pkg/cosign/verify.go#L907

hectorj2f replied:
I added support for checking the certificate expiration although I thought we could also have some helper functions in the timestamp-authority to run these validations (meaning the cert and timestamp.Time). wdyt ?

haydentherapper commented on pkg/cosign/verify.go:
```diff
@@ -674,6 +683,15 @@ func verifyInternal(ctx context.Context, sig oci.Signature, h v1.Hash,
 			return bundleVerified, err
 		}
 	}
+	// TODO: Verify TSA only and return the result with the tsa bundle
```
Like what I mentioned above, the timestamp augments the Rekor bundle rather than supplants it. It's a way to fetch a timestamp from another source. I'd like the verification flow to look more like what's been done in verifyblob, but we should do that refactor outside of this PR imo. For now, I'd propose something like:
* Add in TS verification and check certificate expiration in VerifyTSABundle
* If verification fails, exit
* If succeeds, pass a flag to `VerifyBundle` that says "don't check the timestamp" ([here](https://cs.github.com/sigstore/cosign/blob/44a7117e0cee495a163c92e9f347efb2d9b747c7/pkg/cosign/verify.go?q=verifybundle+org%3Asigstore#L907)) since it was already checked during TS verification. And also [here](https://cs.github.com/sigstore/cosign/blob/44a7117e0cee495a163c92e9f347efb2d9b747c7/pkg/cosign/verify.go?q=tlogValidateCertificate+org%3Asigstore#L462) when calling Rekor online, we don't need to check expiration if the TS is valid and expiration passes

I think we should still check Rekor/bundle if a timestamp is provided, unless we add flags to control this behavior to skip a Rekor check.

cc @asraa

asraa replied:
+1 I think if we should still verify the Rekor bundle. Feel free to followup with a flag to control this behavior (--skip-tlog-verify)

hectorj2f replied:
Yes, blob support will be done as part of a different PR.

I'll add that `skip-tlog-verify` flag as part of a follow-up PR.

hectorj2f commented on cmd/cosign/cli/options/verify.go:
```diff
@@ -132,6 +141,9 @@ type VerifyBlobOptions struct {
 	Registry            RegistryOptions
 	CommonVerifyOptions CommonVerifyOptions
+
+	TSAServerURL     string
```
I was doubting to add these flags without supporting its functionality for the other verify commands.

priyawadhwa replied:
i'm pretty sure all verify commands end up calling `verifyInternal` where the actual TSA verification happens, so adding in the flags should just work!

hectorj2f commented on internal/pkg/cosign/tsa/signer.go:
```diff
@@ -0,0 +1,137 @@
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("Calling TSA authority ...")
```
I removed it. It looks bad when verifying multiple signatures for one image.

hectorj2f commented:
@haydentherapper @priyawadhwa I've addressed all your comments. Please, have a look.

On the other hand I have a list of actions items to tackle in a different PR https://github.com/sigstore/cosign/pull/2460#discussion_r1023190745.

haydentherapper commented on pkg/cosign/verify.go:
```diff
@@ -928,6 +947,58 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, rekorClient *client.Re
+	if err != nil {
+		return false, fmt.Errorf("error reading certification chain path file: %w", err)
+	}
+	certPool := x509.NewCertPool()
```
Can you add a TODO to make sure we update this logic once we finish https://github.com/sigstore/timestamp-authority/issues/121, because this isn't a sufficient check (since this sets the leaf as a the "trusted" cert)

hectorj2f replied:
Done

haydentherapper commented on pkg/cosign/verify.go:
```diff
@@ -928,6 +947,58 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, rekorClient *client.Re
+	if err != nil {
+		return false, fmt.Errorf("reading DecodeString: %w", err)
+	}
+	pemBytes, err := os.ReadFile(filepath.Clean(tsaCertChainPath))
```
Can you have a TODO to support certs from TUF? Just to make sure we come back to this once we spin up a timestamp server

hectorj2f replied:
Done

haydentherapper reviewed (commented):
LGTM

priyawadhwa reviewed (approved):
nice!
